grpc/federation/cel/url.go

```diff
@@ -0,0 +1,611 @@
+					if err != nil {
+						return types.NewErr(err.Error())
+					}
+					return types.String(v.RequestURI())
+				},
+			),
+		),
+		BindMemberFunction(
+			"resolveReference",
+			MemberOverloadFunc(createURLID("resolveReference_url_url_url"), URLType, []*cel.Type{URLType}, URLType,
+				func(_ context.Context, self ref.Val, args ...ref.Val) ref.Val {
+					v, err := self.Value().(*URL).GoURL()
+					if err != nil {
+						return types.NewErr(err.Error())
+					}
+
+					r, err := args[0].Value().(*URL).GoURL()
+					if err != nil {
+						return types.NewErr(err.Error())
+					}
+					u := v.ResolveReference(&r)
+
+					return lib.toURLValue(*u)
+				},
+			),
+		),
+		BindMemberFunction(
+			"string",
+			MemberOverloadFunc(createURLID("string_url_string"), URLType, []*cel.Type{}, cel.StringType,
+				func(_ context.Context, self ref.Val, args ...ref.Val) ref.Val {
+					v, err := self.Value().(*URL).GoURL()
+					if err != nil {
+						return types.NewErr(err.Error())
+					}
+					return types.String(v.String())
+				},
+			),
+		),
+		BindMemberFunction(
+			"unmarshalBinary",
+			MemberOverloadFunc(createURLID("unmarshalBinary_url_bytes_url"), URLType, []*cel.Type{cel.BytesType}, URLType,
+				func(_ context.Context, self ref.Val, args ...ref.Val) ref.Val {
+					var u url.URL
+					err := u.UnmarshalBinary(args[0].(types.Bytes))
+					if err != nil {
+						return types.NewErr(err.Error())
+					}
+					return lib.toURLValue(u)
+				},
+			),
+		),
+
+		// Userinfo functions
+		BindFunction(
+			createURLName("user"),
+			OverloadFunc(createURLID("user_string_userinfo"), []*cel.Type{cel.StringType}, UserinfoType,
+				func(_ context.Context, args ...ref.Val) ref.Val {
+					return lib.toUserinfoValue(string(args[0].(types.String)), "", false)
+				},
+			),
+		),
+		BindFunction(
+			createURLName("userPassword"),
+			OverloadFunc(createURLID("userPassword_string_string_userinfo"), []*cel.Type{cel.StringType, cel.StringType}, UserinfoType,
+				func(_ context.Context, args ...ref.Val) ref.Val {
+					return lib.toUserinfoValue(string(args[0].(types.String)), string(args[1].(types.String)), true)
+				},
+			),
+		),
+		BindMemberFunction(
+			"username",
+			MemberOverloadFunc(createURLID("username_userinfo_string"), UserinfoType, []*cel.Type{}, cel.StringType,
+				func(_ context.Context, self ref.Val, args ...ref.Val) ref.Val {
+					v, err := self.Value().(*Userinfo).GoUserinfo()
+					if err != nil {
+						return types.NewErr(err.Error())
+					}
+					return types.String(v.Username())
+				},
+			),
+		),
+		BindMemberFunction(
+			"password",
+			MemberOverloadFunc(createURLID("password_userinfo_string"), UserinfoType, []*cel.Type{}, cel.StringType,
+				func(_ context.Context, self ref.Val, args ...ref.Val) ref.Val {
+					v, err := self.Value().(*Userinfo).GoUserinfo()
+					if err != nil {
+						return types.NewErr(err.Error())
+					}
+					password, hasPassword := v.Password()
+					if hasPassword {
```

Comment:
nit, Technically, this is different from the original implementation. The original implementation returns `"", false` only when u (*UserInfo) is nil. It does not check whether hasPassword is true or not.

Reply:
It has been fixed.